improver/precipitation_type/snow_splitter.py

```diff
@@ -0,0 +1,152 @@
+# ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+# POSSIBILITY OF SUCH DAMAGE.
+"""module to separate snow and rain contributions from a precipitation diagnostic"""
+
+from typing import Tuple
+
+import numpy as np
+from iris import Constraint
+from iris.cube import Cube, CubeList
+
+from improver import BasePlugin
+from improver.cube_combiner import Combine
+from improver.utilities.cube_checker import assert_spatial_coords_match
+
+
+class SnowSplitter(BasePlugin):
+    """A class to separate the contribution of rain or snow from the precipitation
+    rate/accumulation. This is calculated using the probability of rain and snow
+    at the surface to determine what fraction of the precipitation rate/accumulation
+    is rain or snow.
+     """
+
+    def __init__(self, variable: str):
+        """
+        Sets up Class
+
+        Args:
+            variable:
+                A string of either rain or snow depending on which diagnostic's
+                rate/accumulation should be outputted
+        """
+        self.variable = variable
+
+    @staticmethod
+    def separate_input_cubes(cubes: CubeList) -> Tuple[Cube, Cube, Cube]:
+        """Separate the input cubelist into cubes of rain and snow.
+
+        Args:
+            cubes:
+                containing:
+                    rain_cube:
+                        Cube of the probability of rain at the surface.
+                    snow_cube:
+                        Cube of the probability of snow at the surface.
+                    precip_cube:
+                        Cube of either precipitation rate or precipitation accumulation.
+
+        Returns:
+            A tuple of rain_cube,snow_cube and precip_cube in that order.
+        """
+
+        rain_cube = cubes.extract_cube(
+            Constraint(cube_func=lambda cube: "rain" in cube.name())
+        )
+        snow_cube = cubes.extract_cube(
+            Constraint(cube_func=lambda cube: "snow" in cube.name())
+        )
+        precip_cube = cubes.extract_cube(
+            Constraint(cube_func=lambda cube: "precipitation" in cube.name())
+        )
+
+        return (rain_cube, snow_cube, precip_cube)
+
+    def process(self, cubes: CubeList,) -> Cube:
+        """
+        Splits the precipitation cube data into a snow or rain contribution.
+
+        Whether the output is a rate or accumulation will depend on the precipitation_cube.
+        self.variable will determine whether the outputted cube is a cube of snow or rain.
+
+        The probability of rain and snow at the surfaces should only contain 1's where the
+        precip type is present at the surface and0's where the precip type is not present
+        at the surface. These cubes need to be consistent with eachother such that either
+        rain or snow is always present at the surface (e.g. at no grid square can both
+        diagnostics have a probability of 0).
```

Comment:
I think we need to test for this case and raise an exception if it happens. The maths you've put in would result in 50-50 split when both probs are zero (a good outcome, but I think it shouldn't do this silently).

Reply:
Done